File: app/api/aica_api/services/scenario_registry.py
"""Scenario registry (T019) — scans scenarios_dir and loads ScenarioDefs.

Scans the scenarios directory for ``*.json`` files.  Valid definitions are
loaded and indexed by id; invalid ones are reported in an errors list.

Public API:
  ScenarioRegistry(scenarios_dir: Path)
    .list_summaries() -> list[dict]   — id, version, type, review_focus
    .list_errors()    -> list[dict]   — {file, error} for invalid files
    .get(scenario_id) -> ScenarioDef | None
    .is_compatible(scenario, package) -> bool
"""

from __future__ import annotations

import pathlib
from typing import Any

from pydantic import ValidationError

from aica_api.models.package import PackageManifest
from aica_api.models.scenario import ScenarioDef
from aica_api.storage.file_store import read_json


class ScenarioRegistry:
    """In-memory index of all available scenarios in a scenarios directory."""

    def __init__(self, scenarios_dir: pathlib.Path) -> None:
        self._scenarios: dict[str, ScenarioDef] = {}
        self._errors: list[dict[str, Any]] = []
        self._scan(scenarios_dir)

    # ── Public API ─────────────────────────────────────────────────────────

    def list_summaries(self) -> list[dict]:
        """Return lightweight summaries for all successfully loaded scenarios."""
        return [
            {
                "id": sc.id,
                "version": sc.version,
                "type": sc.type,
                "review_focus": sc.review_focus,
                "total_duration_seconds": sc.total_duration_seconds,
                "tick_seconds": sc.tick_seconds,
            }
            for sc in self._scenarios.values()
        ]

    def list_errors(self) -> list[dict]:
        """Return error entries for scenarios that failed to load."""
        return list(self._errors)

    def get(self, scenario_id: str) -> ScenarioDef | None:
        """Return the full ScenarioDef for a scenario_id, or None."""
        return self._scenarios.get(scenario_id)

    def is_compatible(
        self, scenario: ScenarioDef, package: PackageManifest
    ) -> bool:
        """Return True if the scenario type is in the package's compat list."""
        return scenario.type in package.compatible_scenario_types

    # ── Private helpers ────────────────────────────────────────────────────

    def _scan(self, scenarios_dir: pathlib.Path) -> None:
        """Scan scenarios_dir for .json files."""
        if not scenarios_dir.exists():
            return

        for sc_file in sorted(scenarios_dir.glob("*.json")):
            if not sc_file.is_file():
                continue
            try:
                data = read_json(str(sc_file))
                scenario = ScenarioDef(**data)
                self._scenarios[scenario.id] = scenario
            except (ValidationError, Exception) as exc:
                self._errors.append(
                    {
                        "file": str(sc_file),
                        "error": str(exc),
                    }
                )

File: htmlapp/frontend/src/engine/services/scenario_registry.ts
/**
 * Scenario registry — in-memory index built from IndexedDB scenario records.
 *
 * Ported from `app/api/aica_api/services/scenario_registry.py` (behavior-of-record).
 * The Python registry scans a `scenarios_dir` for `*.json` files at
 * construction time. The htmlapp offline build has no filesystem: scenario
 * definitions are bundled JSON (`src/data/scenarios/`) and seeded into
 * IndexedDB by `storage/db.ts#seedDefaults()` on first launch. This registry
 * reads the already-seeded `ScenarioRecord` rows via `scenarios_store`
 * instead of scanning a directory.
 *
 * Because bundled scenarios are validated at build time, `listSummaries()`
 * always returns an empty `errors` array for now (mirrors `list_errors()`
 * shape; a later slice that adds user-supplied scenario import will
 * populate it).
 *
 * Public API mirrors the Python registry:
 *   scenarioRegistry.listSummaries() -> Promise<{ scenarios, errors }>
 *   scenarioRegistry.get(id)         -> Promise<ScenarioDef>
 *   scenarioRegistry.isCompatible(scenario, package) -> boolean
 */
import { scenariosStore } from '../../storage/scenarios_store'
import type { PackageManifest, RegistryError, ScenarioDef, ScenarioSummary } from '../../api/types'

function personaLabel(persona: Record<string, unknown>): string {
  const name = (persona as { name?: unknown }).name
  return typeof name === 'string' ? name : ''
}

export const scenarioRegistry = {
  /** Return lightweight summaries for all seeded scenarios, plus any load errors. */
  async listSummaries(): Promise<{ scenarios: ScenarioSummary[]; errors: RegistryError[] }> {
    const records = await scenariosStore.list()
    const scenarios: ScenarioSummary[] = records.map((rec) => ({
      id: rec.def.id,
      version: rec.def.version,
      type: rec.def.type,
      persona_label: personaLabel(rec.def.persona),
      review_focus: rec.def.review_focus,
    }))
    return { scenarios, errors: [] }
  },

  /** Return the full ScenarioDef for a scenario_id. Throws if not found. */
  async get(id: string): Promise<ScenarioDef> {
    const rec = await scenariosStore.get(id)
    if (!rec) {
      throw new Error(`Scenario '${id}' not found or invalid`)
    }
    return rec.def
  },

  /** Return true if the scenario type is in the package's compat list. */
  isCompatible(scenario: ScenarioDef, pkg: PackageManifest): boolean {
    return pkg.compatible_scenario_types.includes(scenario.type)
  },
}
